gps/vcs_repo.go

```diff
@@ -103,13 +104,20 @@ func (r *gitRepo) get(ctx context.Context) error {
 }
 
 func (r *gitRepo) fetch(ctx context.Context) error {
+
+	// Hash the remote location to generate the name
+	// of the local refs/pulls namespace.
+	h := md5.New()
+	h.Write([]byte(r.RemoteLocation))
+
 	cmd := commandContext(
 		ctx,
 		"git",
 		"fetch",
 		"--tags",
 		"--prune",
 		r.RemoteLocation,
+		fmt.Sprintf("+refs/pull/*:refs/pull/%x/*", h.Sum(nil)),
```

Comment:
If refs/pull doesn't exist, because the git provider is not github, will this hurt anything? Or just be ignored?

Reply:
Hi @carolynvs,

I've no clue. I didn't test it and didn't want to in case it didn't work :)

However, I will test this and update the PR if necessary.

Reply:
Hi @carolynvs,

Good news everyone!

```shell
$ git fetch -v --progress origin +refs/notreal/*:refs/notreal/origin/* && echo $?
0
```

It appears if the ref structure does not exist, Git simply ignores it.

Reply:
Hi @carolynvs,

Just to prove it, here it is again with the `pull` ref spec:

```shell
$ git fetch -v --progress origin +refs/pull/*:refs/pull/origin/* && echo $?
POST git-upload-pack (gzip 4377 to 2270 bytes)
remote: Counting objects: 131, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 131 (delta 76), reused 76 (delta 76), pack-reused 52
Receiving objects: 100% (131/131), 49.20 KiB | 0 bytes/s, done.
Resolving deltas: 100% (77/77), completed with 66 local objects.
From https://github.com/thecodeteam/gocsi
 * [new ref]         refs/pull/10/head  -> refs/pull/origin/10/head
 * [new ref]         refs/pull/10/merge -> refs/pull/origin/10/merge
 * [new ref]         refs/pull/11/head  -> refs/pull/origin/11/head
 * [new ref]         refs/pull/12/head  -> refs/pull/origin/12/head
 * [new ref]         refs/pull/13/head  -> refs/pull/origin/13/head
 * [new ref]         refs/pull/14/head  -> refs/pull/origin/14/head
 * [new ref]         refs/pull/15/head  -> refs/pull/origin/15/head
 * [new ref]         refs/pull/16/head  -> refs/pull/origin/16/head
 * [new ref]         refs/pull/17/head  -> refs/pull/origin/17/head
 * [new ref]         refs/pull/18/head  -> refs/pull/origin/18/head
 * [new ref]         refs/pull/19/head  -> refs/pull/origin/19/head
 * [new ref]         refs/pull/2/head   -> refs/pull/origin/2/head
 * [new ref]         refs/pull/20/head  -> refs/pull/origin/20/head
 * [new ref]         refs/pull/20/merge -> refs/pull/origin/20/merge
 * [new ref]         refs/pull/21/head  -> refs/pull/origin/21/head
 * [new ref]         refs/pull/23/head  -> refs/pull/origin/23/head
 * [new ref]         refs/pull/23/merge -> refs/pull/origin/23/merge
 * [new ref]         refs/pull/27/head  -> refs/pull/origin/27/head
 * [new ref]         refs/pull/28/head  -> refs/pull/origin/28/head
 * [new ref]         refs/pull/29/head  -> refs/pull/origin/29/head
 * [new ref]         refs/pull/3/head   -> refs/pull/origin/3/head
 * [new ref]         refs/pull/30/head  -> refs/pull/origin/30/head
 * [new ref]         refs/pull/31/head  -> refs/pull/origin/31/head
 * [new ref]         refs/pull/32/head  -> refs/pull/origin/32/head
 * [new ref]         refs/pull/35/head  -> refs/pull/origin/35/head
 * [new ref]         refs/pull/36/head  -> refs/pull/origin/36/head
 * [new ref]         refs/pull/37/head  -> refs/pull/origin/37/head
 * [new ref]         refs/pull/38/head  -> refs/pull/origin/38/head
 * [new ref]         refs/pull/4/head   -> refs/pull/origin/4/head
 * [new ref]         refs/pull/40/head  -> refs/pull/origin/40/head
 * [new ref]         refs/pull/40/merge -> refs/pull/origin/40/merge
 * [new ref]         refs/pull/41/head  -> refs/pull/origin/41/head
 * [new ref]         refs/pull/42/head  -> refs/pull/origin/42/head
 * [new ref]         refs/pull/43/head  -> refs/pull/origin/43/head
 * [new ref]         refs/pull/44/head  -> refs/pull/origin/44/head
 * [new ref]         refs/pull/45/head  -> refs/pull/origin/45/head
 * [new ref]         refs/pull/46/head  -> refs/pull/origin/46/head
 * [new ref]         refs/pull/47/head  -> refs/pull/origin/47/head
 * [new ref]         refs/pull/48/head  -> refs/pull/origin/48/head
 * [new ref]         refs/pull/49/head  -> refs/pull/origin/49/head
 * [new ref]         refs/pull/5/head   -> refs/pull/origin/5/head
 * [new ref]         refs/pull/50/head  -> refs/pull/origin/50/head
 * [new ref]         refs/pull/51/head  -> refs/pull/origin/51/head
 * [new ref]         refs/pull/52/head  -> refs/pull/origin/52/head
 * [new ref]         refs/pull/53/head  -> refs/pull/origin/53/head
 * [new ref]         refs/pull/54/head  -> refs/pull/origin/54/head
 * [new ref]         refs/pull/55/head  -> refs/pull/origin/55/head
 * [new ref]         refs/pull/56/head  -> refs/pull/origin/56/head
 * [new ref]         refs/pull/57/head  -> refs/pull/origin/57/head
 * [new ref]         refs/pull/57/merge -> refs/pull/origin/57/merge
 * [new ref]         refs/pull/58/head  -> refs/pull/origin/58/head
 * [new ref]         refs/pull/59/head  -> refs/pull/origin/59/head
 * [new ref]         refs/pull/59/merge -> refs/pull/origin/59/merge
 * [new ref]         refs/pull/6/head   -> refs/pull/origin/6/head
 * [new ref]         refs/pull/65/head  -> refs/pull/origin/65/head
 * [new ref]         refs/pull/66/head  -> refs/pull/origin/66/head
 * [new ref]         refs/pull/67/head  -> refs/pull/origin/67/head
 * [new ref]         refs/pull/68/head  -> refs/pull/origin/68/head
 * [new ref]         refs/pull/69/head  -> refs/pull/origin/69/head
 * [new ref]         refs/pull/7/head   -> refs/pull/origin/7/head
 * [new ref]         refs/pull/70/head  -> refs/pull/origin/70/head
 * [new ref]         refs/pull/71/head  -> refs/pull/origin/71/head
 * [new ref]         refs/pull/71/merge -> refs/pull/origin/71/merge
 * [new ref]         refs/pull/72/head  -> refs/pull/origin/72/head
 * [new ref]         refs/pull/73/head  -> refs/pull/origin/73/head
 * [new ref]         refs/pull/74/head  -> refs/pull/origin/74/head
 * [new ref]         refs/pull/74/merge -> refs/pull/origin/74/merge
 * [new ref]         refs/pull/75/head  -> refs/pull/origin/75/head
 * [new ref]         refs/pull/76/head  -> refs/pull/origin/76/head
 * [new ref]         refs/pull/77/head  -> refs/pull/origin/77/head
 * [new ref]         refs/pull/78/head  -> refs/pull/origin/78/head
 * [new ref]         refs/pull/79/head  -> refs/pull/origin/79/head
 * [new ref]         refs/pull/8/head   -> refs/pull/origin/8/head
 * [new ref]         refs/pull/80/head  -> refs/pull/origin/80/head
 * [new ref]         refs/pull/81/head  -> refs/pull/origin/81/head
 * [new ref]         refs/pull/82/head  -> refs/pull/origin/82/head
 * [new ref]         refs/pull/83/head  -> refs/pull/origin/83/head
 * [new ref]         refs/pull/85/head  -> refs/pull/origin/85/head
 * [new ref]         refs/pull/9/head   -> refs/pull/origin/9/head
0
```

Reply:
Thanks for trying that out!

I have a bug right now that needs this, I'll give it a go and report back.

Reply:
@akutz I'm not quite sure what I should be putting into my Gopkg.toml in order to say that I want a pull request. Is it "pull/origin/X/head"? Some hash?

Reply:
Hi @carolynvs,

I'm blocked on this as well.

> I'm not quite sure what I should be putting into my Gopkg.toml in order to say that I want a pull request. Is it "pull/origin/X/head"? Some hash?

The source is the URI of the repo to which the PR was submitted. Then you just set `revision = "COMMIT_ID"` like any other commit revision. You can figure out the commit ID of a PR commit by looking at the Travis-CI build log or just clone your PR locally and find the head commit ID: `git fetch +refs/pull/*:refs/pull/REMOTE_NAME/*` and then do a reflog on that ref spec.

Reply:
Hi @carolynvs,

See thecodeteam/gocsi#71 for my PR that uses this and [this part](https://github.com/akutz/gocsi/blob/bab7c239780744b923dc6122faa6fd3ce84737f0/gocsi.sh#L314-L341) of the script that uses those variables:

```bash
# Refer to https://github.com/golang/dep/blob/master/docs/Gopkg.toml.md
# for detailed Gopkg.toml documentation.
#
# Refer to https://github.com/toml-lang/toml for detailed TOML docs.
[[constraint]]
  name = "github.com/container-storage-interface/spec"
  branch = "master"
[[constraint]]
  name = "github.com/thecodeteam/gocsi"
EOF

    if [ "$GOCSI_DEP_BRANCH" != "" ]; then
      echo "  branch = \"$GOCSI_DEP_BRANCH\"" >> Gopkg.toml
    elif [ "$GOCSI_DEP_REVISION" != "" ]; then
      echo "  revision = \"$GOCSI_DEP_REVISION\"" >> Gopkg.toml
    elif [ "$GOCSI_DEP_VERSION" != "" ]; then
      echo "  version = \"$GOCSI_DEP_VERSION\"" >> Gopkg.toml
    else
      echo "  branch = \"master\"" >> Gopkg.toml
    fi

    if [ "$GOCSI_DEP_SOURCE" != "" ]; then
      echo "  source = \"$GOCSI_DEP_SOURCE\"" >> Gopkg.toml
    fi

  fi
```

Reply:
Hi @carolynvs,

> Is it "pull/origin/X/head"?

That's an issue. @sdboyer or someone made it such that the `source` field is pattern matched to match a Go package. You cannot put anything interesting as `source` other than `(https?://)?[\w\d\.]+` (or something like that, maybe SSH/git works, I haven't tried). It **does** exclude file URIs, non-standard URIs, etc. It's frustrating. See #1581 and #1583.